### ISSUE TITLE
TypeError when hydration is enabled (hostElement.hasAttribute is not a function)

### ISSUE THREAD

**Issue by phihochzwei** [angular/angular#50133]
### Which @angular/* package(s) are the source of the bug?

Don't known / other

### Is this a regression?

No

### Description

When adding provideClientHydration() to my app.module.ts, the render()-Method breaks with the error

### Please provide the exception or error you saw

```true
TypeError: hostElement.hasAttribute is not a function
    at annotateForHydration (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:2881506)
    at /Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:3491908
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:3748394)
    at _next (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:3748698)
    at ZoneTask2.callback (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:1902313)
    at _ZoneDelegate2.invokeTask (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:1895862)
    at Zone3.runTask (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:1888322)
    at ZoneTask2.invokeTask (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:1897459)
    at ZoneTask2.invoke (/Users/jko13/GIT/storefront/dist/storefront/server/main.js:1:1897256)
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
_                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 16.0.0
Node: 18.10.0
Package Manager: npm 8.19.2
OS: darwin arm64

Angular: 16.0.0
... animations, cdk, cli, common, compiler, compiler-cli, core
... forms, language-service, material, platform-browser
... platform-browser-dynamic, platform-server, router
... service-worker

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.0
@angular-devkit/build-angular   16.0.0
@angular-devkit/core            16.0.0
@angular-devkit/schematics      16.0.0
@angular/google-maps            15.2.9
@nguniversal/builders           16.0.0
@nguniversal/express-engine     16.0.0
@schematics/angular             16.0.0
ng-packagr                      16.0.0
rxjs                            7.5.7
typescript                      5.0.4
webpack                         5.82.0
```

### Anything else?

I am close to tears since if was looking so much forward for this :(

**Comment by JeanMeche** [comment]
Hi, 

Thank you for reporting an issue. Could you provide a reproduction on stackblitz to help us investigate the issue ?

**Comment by AndrewKushnir** [comment]
@phihochzwei thanks for reporting an issue. It'd be very helpful if you could try to narrow down the problem to see which pattern is causing this issue. Once you have a pattern that is causing the problem, we can perform further investigation and help to find a solution. To speed up the investigation, it'd be great to have a minimal repro (in a form of a GitHub repository) @JeanMeche mentioned above. Note: do **not** share the code of your application, try to create a new `ng new` application and use a pattern from your application that is causing the problem.

**Comment by AndrewKushnir** [comment]
@phihochzwei as a temporary solution, if the problem is isolated to one of the components, you can try to opt that component out of hydration using the `ngSkipHydration` attribute as described in [hydration guide](https://angular.io/guide/hydration#how-to-skip-hydration-for-particular-components).

**Comment by phihochzwei** [comment]
Hi @AndrewKushnir  

thanks for the answer. Actually, narrowing it down is my biggest problem here, since the project is quite big. What I found "so far" was, that within ALL OF THE CODE within the whole project ( I even scanned node_modules), the only place where hostElement.hasAttribute gets used seems to be within material. But I will try my best.

**Comment by AndrewKushnir** [comment]
@phihochzwei right, the error is coming from this code:

https://github.com/angular/angular/blob/b3730081c3ddc7726c6fb1272e97ee6862d3b3a4/packages/core/src/hydration/annotate.ts#L111-L113

and the `hostElement` at that moment is expected to be a element node (which is a component host node). Based on the error message, it looks like it's either a comment node, a text node or some other object. We'd need to understand what this object is and under which circumstances this happens.

As a next step in the investigation, you can try updating the code in the `node_modules` or in the generated `main.js` that is used to run the app on the server and output some information into a console: like `hostElement.nodeType` and `hostElement.textContent` (if it's a DOM node) and may be try to stringify the value (for ex. using `JSON.stringify`) if it's not a DOM node.

Please let us know what you find.

**Comment by phihochzwei** [comment]
@AndrewKushnir  Thank you all so much for your help and time. 

I added some console.logs and it seems, that hostElement at this point is not an element node but an array WITH element nodes (according to the console log 125). When I access the first element of this array, it returns me my app.component.

**Comment by phihochzwei** [comment]
In addition: If I change the code to 

...hostElement. && !Array.isArray(hostElement) && hostElement.hasAttribut...  the page gets rendered. But my uneducated guess would be, that clientHydration is then completely skipped

**Comment by AndrewKushnir** [comment]
@phihochzwei thanks for additional information. It'd be very helpful to understand why the `hostElement` is an array at that point. It looks like this is your root component, so I'm curious which bootstrapping method do you use for your application?

**Comment by phihochzwei** [comment]
@AndrewKushnir  Currently, this one:

`platformBrowserDynamic()
		.bootstrapModule(AppModule)
		.then((module) => {
			if (!environment.production) {
				enableDebugTools(module.injector.get(ApplicationRef).components[0]);
			}
		})

		.catch((err) => console.error(err));`

**Comment by AndrewKushnir** [comment]
@phihochzwei a couple additional questions:
- How big is the array? Does it contain 1-2 elements in it or it's more like 20+ elements?
- A couple lines above that code you should have an access to the `lView` variable. Could you please also do `console.log(lView[2] & 256)`? This should indicate whether the view is destroyed or not.
- Is there any change if you comment out the `enableDebugTools` line?

Thank you.

**Comment by phihochzwei** [comment]
@AndrewKushnir 
- The array contains 125 elements
- console.log(lView[2] & 256) returns "0"
- I removed the enableDebugTools

Thank YOU :)

**Comment by AndrewKushnir** [comment]
@phihochzwei thanks for additional information 👍 

I was able to reproduce the error and there is a fix in progress in PR https://github.com/angular/angular/pull/50136.

Could you please try to test the behavior of your application with this fix applied? You can update the `package.json` file locally and replace the `@angular/core` dependency with the following package:

```
  ...
  "@angular/core": "https://output.circle-artifacts.com/output/job/54cbbe34-29e4-48e4-b39e-93d43c25e84e/artifacts/0/angular/core-pr50136-fe64b8fe12.tgz",
  ...
```

After that you can run `npm install --force` and test the behavior of the app once again.

Please let us know if that resolved the issue.

Thank you.

**Comment by phihochzwei** [comment]
I will try right away

**Comment by phihochzwei** [comment]
@AndrewKushnir 

That seemed to do the trick :) The server renders. You are lifesaver <3 

However, two things bother me:

- In the docs I read something about hydration information outputted in the console, I don't see anything like that. Does that mean, that it is not hydrating? 

- I receive two errors in the console:

`TypeError: Cannot set properties of null (setting '__ngContext__')
    at Wr (vendor.9826d529f0d1a225.js:1:205395)
    at Object.dp (vendor.9826d529f0d1a225.js:1:250040)
    at template (main.f5c56b123be9d080.js:1:9799)
    at xh (vendor.9826d529f0d1a225.js:1:230464)
    at wl (vendor.9826d529f0d1a225.js:1:238006)
    at vE (vendor.9826d529f0d1a225.js:1:237894)
    at EE (vendor.9826d529f0d1a225.js:1:238210)
    at wl (vendor.9826d529f0d1a225.js:1:238221)
    at Ba.create (vendor.9826d529f0d1a225.js:1:246021)
    at e.bootstrap (vendor.9826d529f0d1a225.js:1:293052)
    at vendor.9826d529f0d1a225.js:1:291763
    at Array.forEach (<anonymous>)
    at e._moduleDoBootstrap (vendor.9826d529f0d1a225.js:1:291750)
    at vendor.9826d529f0d1a225.js:1:291452
    at v.invoke (polyfills.531004ca9e7bbe37.js:1:6751)
    at Object.onInvoke (vendor.9826d529f0d1a225.js:1:283999)
    at v.invoke (polyfills.531004ca9e7bbe37.js:1:6691)
    at M.run (polyfills.531004ca9e7bbe37.js:1:2117)
    at polyfills.531004ca9e7bbe37.js:1:16961
    at v.invokeTask (polyfills.531004ca9e7bbe37.js:1:7369)
    at Object.onInvokeTask (vendor.9826d529f0d1a225.js:1:283815)
    at v.invokeTask (polyfills.531004ca9e7bbe37.js:1:7290)
    at M.runTask (polyfills.531004ca9e7bbe37.js:1:2734)
    at _ (polyfills.531004ca9e7bbe37.js:1:9386)
    at invokeTask (polyfills.531004ca9e7bbe37.js:1:8450)
    at invoke (polyfills.531004ca9e7bbe37.js:1:8321)
    at p.args.<computed> (polyfills.531004ca9e7bbe37.js:1:28009)
(`

is this somehow related or is this something different?

**Comment by AndrewKushnir** [comment]
Thanks for checking the fix!

> In the docs I read something about hydration information outputted in the console, I don't see anything like that. Does that mean, that it is not hydrating?

The message will only be displayed in the dev mode. If you build in prod mode, it will not be there.

> I receive two errors in the console:

Could you please try to build the app in dev mode (it looks like the error message is from the prod mode) and share the error message that you see? If the error is coming from one of the components, you can try applying the `ngSkipHydration` attribute to it.

**Comment by phihochzwei** [comment]
Okay, THIS looks promising :) <3

`Angular hydrated 112 component(s) and 1844 node(s)`

It would be REALLY nice, if this message could be turned on for "production light" (staging systems)

So, would can I do about this error in Production mode ?

**Comment by AndrewKushnir** [comment]
> It would be REALLY nice, if this message could be turned on for "production light" (staging systems)

Could you please create a new ticket (a feature request) about this? This would help us to collect feedback from the community.

> So, what can I do about this error in Production mode ?
> I checked the source of the error (the main.js-part) and added ngSkipHydration to the component that is referred there, but this had no effect
> h.YNc(5, Q, 1, 0, "deich-lib-header", 2), (Just in case this helps)

Does the application work in the dev mode, but fails in the prod mode? If the same problem happens in dev mode, there should be a more detailed error message and a better stack trace.

**Comment by phihochzwei** [comment]
No, everything Fine in devmode, no errors

**Comment by AndrewKushnir** [comment]
> No, everything Fine in devmode, no errors

That's unexpected, the main logic should be the same in dev and prod modes.

Do you perform any changes/optimizations to the server-generated HTML in prod mode? In some applications we've seen HTML optimization/transformation which remove whitespaces and comment nodes. Altering HTML generated by the server might cause issues during hydration, which expects the DOM structure to be the same. We've also added a note on this in the [hydration guide](https://angular.io/guide/hydration#constraints).

**Comment by phihochzwei** [comment]
In fact, I do. But not only in Prod, it's part of my response. I will deactivate it, just a second

**Comment by phihochzwei** [comment]
My apologies. I can confirm, that my removal of the dead comment nodes caused the problem
`.split('<!---->').join('')` removed and all is working fine :) 

I misinterpreted the hydration guide on that. 

Thank you so much for your time and help!

**Comment by AndrewKushnir** [comment]
@phihochzwei thanks for the update! Happy to hear that everything works as expected now! 🎉 

We will proceed with the fix from PR #50136, it should be available in v16.0.1 release next week.

**Comment by joldibaev** [comment]
I also have the same error due to:
```
constructor(private viewContainerRef: ViewContainerRef) {}
```

**Comment by AndrewKushnir** [comment]
@joldibaev the problem should be resolved in the upcoming 16.0.1 next week. Meanwhile, you can try to test the behavior of your app with the fix using information from https://github.com/angular/angular/issues/50133#issuecomment-1534151814. Thank you.

**Comment by AndrewKushnir** [comment]
Just wanted to share a quick update: Angular v16.0.1 was released today. Angular v16.0.1 includes the fix for this issue, please try updating to v16.0.1 and let us know if he problem still exists. Thank you.
